### PR TITLE
refactor(sql): stabilize binder alias semantics across aggregate, window, and clause rewrites

### PR DESCRIPTION
**Pull request by forsaken628**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

This PR refactors the SQL binder to preserve a stable semantic view of SELECT aliases across WHERE, HAVING, QUALIFY, and ORDER BY, while separating prepass aggregate discovery from final projection rewriting.

  - Introduce aggregate prepass alias/catalog analysis to detect aggregate and window dependencies before clause binding.
  - Keep clause alias resolution based on original SELECT semantics, avoiding accidental dependence on already-rewritten projection state.
  - Refactor projection handling around SelectInfo, so projection, DISTINCT, and ORDER BY share the same source/projection scalar model.
  - Tighten QUALIFY, window, grouping, and aggregate interactions, including UDAF and SRF edge cases.
  - Expand binder semantic coverage with focused golden tests for clauses, grouping, window paths, named windows, and subquery metadata side effects.

## Tests

- [x] Unit Test
- [ ] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [x] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/databendlabs/databend/19618)

### CODE REVIEW

**Comment by forsaken628** [comment]
@codex review

File: src/query/sql/src/planner/binder/aggregate.rs [hunk 1]
```diff
@@ -864,6 +1397,157 @@ impl Binder {
...
+        }
+        for order in order_by {
+            order.expr.drive(&mut visitor);
+        }
+
+        let mut udaf_names = HashSet::new();
+        let tenant = self.ctx.get_tenant();
+        let provider = UserApiProvider::instance();
+        for name in visitor.names {
+            if AggregateFunctionFactory::instance().contains(name.as_str())
```
Comment (b41sh) [inline src/query/sql/src/planner/binder/aggregate.rs]: don't need check `AggregateFunctionFactory` contains the function name, because `is_builtin_function` includes an `AggregateFunctionFactory` check.

File: src/query/sql/src/planner/binder/aggregate.rs [hunk 2]
```diff
@@ -845,6 +857,527 @@ impl<'a> VisitorMut<'a> for AggregateRewriter<'a> {
...
+                    &by_name,
+                    &ast,
+                ),
+                contains_aggregate: Binder::aggregate_prepass_contains_aggregate(
+                    name_resolution_ctx,
+                    functions,
+                    &ast,
+                ),
+                contains_window: Binder::aggregate_prepass_contains_window(&ast),
+                contains_subquery: Binder::aggregate_prepass_contains_subquery(&ast),
```
Comment (b41sh) [inline src/query/sql/src/planner/binder/aggregate.rs]: `collect_referenced_aliases`, `aggregate_prepass_contains_aggregate`, `aggregate_prepass_contains_window` and `aggregate_prepass_contains_subquery` use four `Visitor` to check whether the `ast` contains `aggregate`, `window`, `subquery` and `alias`, we can use one `Visitor` to do this.

File: src/query/sql/src/planner/binder/bind_query/bind_select.rs [hunk 3]
```diff
@@ -54,8 +54,160 @@ use crate::optimizer::ir::SExpr;
...
+#[derive(Clone, Default)]
+struct SelectClauseFacts {
+    having: Option<SelectClauseFact>,
+    qualify: Option<SelectClauseFact>,
+    order_by: Vec<SelectClauseFact>,
+}
+
+impl SelectClauseFacts {
+    fn iter_aggregate_prepass(&self) -> impl Iterator<Item = &SelectClauseFact> {
+        self.having.iter().chain(self.order_by.iter())
```
Comment (b41sh) [inline src/query/sql/src/planner/binder/bind_query/bind_select.rs]: `qualify` is ignored here.

File: src/query/sql/src/planner/binder/aggregate.rs [hunk 4]
```diff
@@ -845,6 +857,527 @@ impl<'a> VisitorMut<'a> for AggregateRewriter<'a> {
...
+            expr_context: self.expr_context,
+            source: self
+                .expansion_stack
+                .last()
+                .cloned()
+                .map(AggregatePrepassSource::AliasExpansion)
+                .unwrap_or(AggregatePrepassSource::DirectClause),
+            expr: expr.clone(),
+            contains_window: Binder::aggregate_prepass_contains_window(expr),
+            contains_subquery,
```
Comment (b41sh) [inline src/query/sql/src/planner/binder/aggregate.rs]: `contains_subquery` is not needed, because if `contains_subquery` is `true`, we will not add this fragment.

File: src/query/sql/src/planner/binder/bind_query/bind_select.rs [hunk 5]
```diff
@@ -142,11 +294,51 @@ impl Binder {
...
+        let aggregate_prepass_facts = self.derive_aggregate_prepass_facts(
+            &aggregate_prepass_functions,
+            &aggregate_prepass_aliases,
+            &clause_facts,
+        );
+        let global_view = SelectGlobalView {
+            semantic_aliases,
+            clause_facts,
+            aggregate_prepass_facts,
+        };
```
Comment (b41sh) [inline src/query/sql/src/planner/binder/bind_query/bind_select.rs:206]: We can add those info to `BindContext`, don't need add a new struct `SelectGlobalView`.

File: src/query/sql/src/planner/binder/bind_query/bind_select.rs [hunk 6]
```diff
@@ -142,11 +294,51 @@ impl Binder {
...
+
+        self.register_aggregate_prepass_facts(
+            &mut from_context,
+            &global_view.semantic_aliases,
+            &global_view.aggregate_prepass_facts,
+        )?;
```
Comment (b41sh) [inline src/query/sql/src/planner/binder/bind_query/bind_select.rs]: Those codes collects alias names, aggregate functions, window functions related info, we can write these codes into a separate file for better reuse in the future.

File: src/query/sql/src/planner/binder/bind_query/bind_select.rs (hunk 5 above)
Comment (forsaken628) [reply src/query/sql/src/planner/binder/bind_query/bind_select.rs:206]: I do not think SelectGlobalView should be merged into BindContext. The key issue is not whether their fields look similar, but that they serve different roles. BindContext is
  the binder's general mutable scope state: it carries the current scope's columns, aggregate state, window state, SRF state, CTE state, parent scope, and similar binder-owned
  data, as defined in src/query/sql/src/planner/binder/bind_context.rs:134. SelectGlobalView, by contrast, exists only inside the local bind_select flow and holds read-only
  intermediate results derived during SELECT semantic analysis: the alias semantic snapshot, clause facts, and aggregate prepass facts, as defined in src/query/sql/src/planner/
  binder/bind_query/bind_select.rs:109.

  More importantly, SelectGlobalView acts as a frozen intermediate state that removes mutable side effects from later binding phases. In bind_select, semantic_aliases is captured
  before later aggregate, window, and SRF rewrites continue mutating from_context and select_list, as shown in src/query/sql/src/planner/binder/bind_query/bind_select.rs:281.
  WHERE, QUALIFY, and ORDER BY then deliberately read from that same stable semantic snapshot instead of whatever mutable state happens to live in BindContext at a later point in
  the pipeline, as shown in src/query/sql/src/planner/binder/bind_query/bind_select.rs:359. That is also consistent with the surrounding API design: aliases are explicitly passed
  into binding/analyze routines rather than stored inside BindContext, for example in src/query/sql/src/planner/binder/select.rs:71, src/query/sql/src/planner/binder/
  qualify.rs:40, and src/query/sql/src/planner/binder/sort.rs:92; even BindContext::resolve_name takes available_aliases as an external input rather than reading them from
  internal state, in src/query/sql/src/planner/binder/bind_context.rs:369.

  There is also a practical cost to merging them: BindContext is cloned and re-parented frequently through clone, with_parent, and replace, as shown in src/query/sql/src/planner/
  binder/bind_context.rs:279. Putting SELECT-local semantic snapshots into that shared context would make the context heavier and, more importantly, create awkward inheritance
  questions across parent/child scopes about whether such SELECT-phase snapshots should propagate at all. Refactoring in this area is also still ongoing, so preserving a clear
  separation between the general mutable binder context and this SELECT-local frozen snapshot keeps future cleanup and iteration easier instead of hard-wiring transient phase
  state into a core shared structure.
